### PR TITLE
Add brackets around IPv6 kube-dns liveness probe addrs

### LINKED ISSUE #56543: kube-dns pod restarts every 2 minutes in IPv6-only cluster
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line: 
>
/kind bug
> /kind feature
/area ipv6
/sig network

**What happened**:
When a Kubernetes cluster is instantiated in IPv6-only mode via kubeadm, using the latest master code base, then the kube-dns pod is restarted approximately every 2 minutes.

Examination of the kube-dns sidecar logs shows that in an IPv6-only cluster, the liveness probes used for kube-dns are set to use the server IP:port values of "::1:53" and "::1:10053". These are invalid, and the correct IP:port strings should use brackets around the IPv6 address, e.g. "[::1]:53" and "[::1]:10053". 

Similarly, examination of the kube-dns deployment YAML configuration shows the liveness probes using IP:port values of "::1:53" and "::1:10053" rather than "[::1]:53" and "[::1]:10053".

**What you expected to happen**:
The kube-dns pod should stay up indefinitely in an IPv6-only cluster.

**How to reproduce it (as minimally and precisely as possible)**:
Use Kubernetes binaries and docker images built from the latest master. Then start up an IPv6-only cluster using 'kubeadm init ...' with the following kubeadm configuration (where unifiedControlPlaneImage points to the hyperkube image built from latest master):
```
apiVersion: kubeadm.k8s.io/v1alpha1
kind: MasterConfiguration
kubernetesVersion: 1.10.0
api:
  advertiseAddress: fd00::100
networking:
  serviceSubnet: fd00:1234::/110
unifiedControlPlaneImage: diverdane/hyperkube-amd64:v1.9_code_freeze
tokenTTL: 0s
nodeName: kube-master
```
After the kubernetes master node comes up, you will see the kube-dns pod restarted every 2 minutes or so.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): v1.10.0-alpha.0.633+02a7c12cbd0dd3-dirty (latest master)
- Cloud provider or hardware configuration: VirtualBox cluster, Ubuntu 16.04.1 host, CentOS 7 guests
- OS (e.g. from /etc/os-release): CentOS 7
- Kernel (e.g. `uname -a`): 3.10.0-514.26.2.el7.x86_64 #1 SMP Tue Jul 4 15:04:05 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by leblancd**
**What this PR does / why we need it**:
Previously, when an IPv6-only Kubernetes cluster was instantiated
via 'kubeadm init ...', the liveness probes that were
configured by kubeadm for kube-dns would use IP:port
strings of "::1:53" and "::1:10053". These IP:port strings should
instead include brackets around the ::1 IPv6 loopback address, e.g.
"[::1]:53" and "[::1]:10053".

This change adds the necessary brackets around the ::1 IPv6 loopback
address.

Without this change, the kube-dns sidecar container interprets the
bracket-less strings as IPv4 IP:port strings that have too many
colons, and the kube-dns pod is restarted about once every 2 minutes.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #56543

/area ipv6
/sig network

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE

```

### CODE REVIEW

File: cmd/kubeadm/app/phases/addons/dns/dns.go
Comment (kad) [inline cmd/kubeadm/app/phases/addons/dns/dns.go]: I'd rather do `dnsProbeAddr = "[" + dnsBindAddr + "]"` and `dnsProbeAddr = dnsBindAddr` instead of repeating constants. Beside that, looks good.

**Comment by pmichali** [comment]
FYI, here is the error message:

```
curl -k http://[fd00:79::2:0:0:1]:10054/healthcheck/kubedns
{"IsOk":false,"LatencySeconds":0,"Err":"dial udp: address ::1:10053: too many colons in address"}
```

**Review by pmichali — APPROVED** [review]
Tested this under Docker-in-Docker setup, and liveness probe is working fine with this change. Other than the cleanup suggested, this looks good.

File: cmd/kubeadm/app/phases/addons/dns/dns.go
Comment (leblancd) [reply cmd/kubeadm/app/phases/addons/dns/dns.go]: Agreed, will change.

Comment (pmichali) [reply cmd/kubeadm/app/phases/addons/dns/dns.go]: Instead, could use net.JoinHostPort() and pass in dnsBindAddr and port number, using that host/port in template.

**Review by pmichali — APPROVED** [review]

File: cmd/kubeadm/app/phases/addons/dns/dns.go
Comment (leblancd) [reply cmd/kubeadm/app/phases/addons/dns/dns.go]: @pmichali : There would be a bit of a tradeoff in using net.JoinHostPort(), in that it would require more variables being passed to the template (since the template uses 2 separate IP:port combinations, one for port 53, and one for port 10053). For example, the variable setup would look (roughly) like this:
```
	dnsBindAddr = "::1"
	kubeDNSEndpoint = net.JoinHostPort(dnsBindAddr, "53")
	dnsMasqEndpoint = net.JoinHostPort(dnsBindAddr, "10053")
```
And the instantiation of the template would look like this:
```
	struct{ ImageRepository, Arch, Version, DNSBindAddr, KubeDNSEndpoint, DNSMasqEndpoint, DNSDomain, DNSProbeType, MasterTaintKey string }{
```
with the template definition being:
```
       - --probe=kubedns,{{ .KubeDNSEndPoint }},kubernetes.default.svc.{{ .DNSDomain }},5,{{ .DNSProbeType }}
       - --probe=dnsmasq,{{ .DNSMasqEndpoint }},kubernetes.default.svc.{{ .DNSDomain }},5,{{ .DNSProbeType }}
```
My preference would be the less verbose option adding square brackets directly (with one template variable for DNS probe definitions):
```
    dnsProbeAddr = "[" + dnsBindAddr + "]"
```

**Comment by danehans** [comment]
This is a bug that will effect IPv6 in the 1.9 release. Is it possible to get this merged for 1.9?

**Review by luxas — APPROVED** [review]
/approve

**Comment by luxas** [comment]
/hold

The kubeadm PR job failed for this PR: https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/56544/pull-kubernetes-e2e-kubeadm-gce/9508/

It _could_ indeed be a flake, but let's retest before merging the PR

/test pull-kubernetes-e2e-kubeadm-gce

**Comment by luxas** [comment]
/hold cancel
It was a flake, https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/56544/pull-kubernetes-e2e-kubeadm-gce/9509/ passed :+1:

**Pull request merged**
